src/java.base/share/native/launcher/main.c

```diff
@@ -278,15 +343,33 @@ main(int argc, char **argv)
         margv = args->elements;
     }
 
-    // Avoid unexpected process completion when checkpointing under docker container run
-    // by creating the main process waiting for children before exit.
-    if (is_checkpoint && 1 == getpid()) {
-        g_child_pid = fork();
-        if (0 < g_child_pid) {
-            // The main process should forward signals to the child.
-            setup_sighandler();
-            const int status = wait_for_children();
-            exit(status);
+    if (is_checkpoint) {
+        const int crac_min_pid_default = 128;
+        const int min_pid = 0 < crac_min_pid ? crac_min_pid : crac_min_pid_default;
+
+        if (getpid() <= min_pid) {
+            // Move PID value for new processes to a desired value
+            // to avoid PID conflicts on restore.
+            {
+                const int res = set_last_pid(min_pid);
+                if (EPERM == res || EACCES == res || EROFS == res) {
+                    spin_last_pid(min_pid);
+                } else if (0 != res) {
+                    errno = res;
+                    perror("set_last_pid");
```

Comment:
This apparently prints a textual version of the error. Better to avoid modifying the global state:

```suggestion
                    fprintf(stderr, "set_last_pid: %s\n", strerror(res))
```